### PR TITLE
DS-4497 Upgrade to Rack 3

### PR DESCRIPTION
**Pull request by ildomar-grings**
## Overview

Upgrade to Rack 3

## What is the change?
- Adjust middleware for Rackup
- Adjust tests specs 
- Bump gems:
  - rack 3.1.4 (was 2.2.9)
  - alchemy-flux 1.6.0 (was 1.2.1)
  - rack-test 2.1.0 (was 1.1.0)

## Notes
The new Rack gem has introduced some changes that directly affect Hoodoo. Changes as such:
- The `rack.input` is no longer rewound after consuming form and multipart data
- Response Headers must be lowercase
- The classes Rack::Server, Rack::Handler and Rack::Lobster have been moved to the rackup gem too and renamed to Rackup::Server, Rackup::Handler and Rackup::Lobster respectively.

Source: https://github.com/rack/rack/blob/main/UPGRADE-GUIDE.md

[DS-4497]: https://loyaltynz.atlassian.net/browse/DS-4497?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/hoodoo/services/middleware/middleware.rb [hunk 1]
```diff
@@ -1160,8 +1150,22 @@ def monkey_log_inbound_request( interaction )
       # Compile the remaining log payload and send it.
 
       unless secure
-        body = interaction.rack_request.body.read( MAXIMUM_LOGGED_PAYLOAD_SIZE )
-               interaction.rack_request.body.rewind()
+        body = interaction.rack_request.body.read( MAXIMUM_LOGGED_PAYLOAD_SIZE ) unless interaction.rack_request.body.nil?
```
Comment (cloudy9101) [inline lib/hoodoo/services/middleware/middleware.rb:1153]: Why `body` might be nil? Why we don't need to care about it before upgrading?

File: lib/hoodoo/services/middleware/middleware.rb [hunk 2]
```diff
@@ -1160,8 +1150,22 @@ def monkey_log_inbound_request( interaction )
...
+
+        # In case the body will get read again later, we need to restore the reader.
+        #
+        # A) When the reader is an Rack::RewindableInput, we can just rewind it.
+        # B) Case not, we need to replace the body with a new reader instance.
+        if body
+          if interaction.rack_request.body.respond_to?(:rewind)
```
Comment (cloudy9101) [inline lib/hoodoo/services/middleware/middleware.rb:1160]: Could we use [rewindable_input](https://github.com/rack/rack/blob/e5801a49ba540de76c7437d256361189bfcdd147/lib/rack/rewindable_input.rb#L24) as we want it rewindable anyway?

**Review by cloudy9101 — APPROVED** [review]
Non blocking comments.

File: lib/hoodoo/services/middleware/middleware.rb [hunk 3]
```diff
@@ -88,14 +88,7 @@ class Middleware
...
     #
-    ALLOWED_QUERIES_LIST = [
-      'offset',
-      'limit',
-      'sort',
-      'direction',
-      'search',
-      'filter'
-    ]
+    ALLOWED_QUERIES_LIST = %w[offset limit sort direction search filter]
```
Comment (davidoram) [inline lib/hoodoo/services/middleware/middleware.rb:91]: This change seems unrelated? Do we need it?

File: lib/hoodoo/services/middleware/middleware.rb [hunk 4]
```diff
@@ -105,10 +98,7 @@ class Middleware
     # entries from the list below could cause their requests to be rejected
     # with a 'platform.malformed' error.
     #
-    ALLOWED_QUERIES_ALL = [
-      '_embed',
-      '_reference'
-    ]
+    ALLOWED_QUERIES_ALL = %w[_embed _reference]
```
Comment (davidoram) [inline lib/hoodoo/services/middleware/middleware.rb:101]: Seems unrelated? Not adding value?

File: lib/hoodoo/services/middleware/middleware.rb (hunk 2 above)
Comment (davidoram) [reply lib/hoodoo/services/middleware/middleware.rb:1160]: Agreed, is it possible to inject the `Rack::RewindableInput::Middleware` so we have unchanged functionality?

File: spec/services/services/response_spec.rb [hunk 5]
```diff
@@ -159,7 +159,7 @@
 
       expected = JSON.generate({})
       expect(status).to eq(200)
-      expect(headers).to eq({'Content-Length' => expected.length.to_s})
+      expect(headers).to eq({'content-length' => expected.length.to_s})
```
Comment (davidoram) [inline spec/services/services/response_spec.rb:162]: Will the fact that headers are now all lower case affect any services that use custom HTTP headers?

**Review by davidoram — COMMENTED** [review]
Commented

File: lib/hoodoo/services/middleware/middleware.rb (hunk 2 above)
Comment (ildomar-grings) [reply lib/hoodoo/services/middleware/middleware.rb:1160]: I tried, and it doesn't work straight away:


https://github.com/LoyaltyNZ/hoodoo/blob/master/spec/spec_helper.rb#L260


```
  Thread.start do
    app = Rack::Builder.new do
      use Rack::RewindableInput::Middleware
      use Hoodoo::Services::Middleware unless app_options[:skip_hoodoo_middleware]
      run app_class.new
    end
```

For some specs this workaround works fine, but a bunch of other ones it fails, raising error 500.

Lemme try to get some more internal details of the cascade problem.

Comment (ildomar-grings) [reply lib/hoodoo/services/middleware/middleware.rb:1160]: Just a single example among hundreds:

https://github.com/LoyaltyNZ/hoodoo/blob/master/spec/services/middleware/middleware_spec.rb#L1224 

```
    describe 'service implementation #show' do
      it 'should get called with correct path data (1)' do

        expect_any_instance_of(RSpecTestServiceStubImplementation).to receive(:show).once do | ignored_rspec_mock_instance, context |
          expect(context.request.uri_path_components).to eq(['12345'])
          expect(context.request.ident).to eq('12345')
          expect(context.request.uri_path_extension).to eq('tar.gz')
        end

        get '/v2/rspec_test_service_stub/12345.tar.gz', nil, { 'CONTENT_TYPE' => 'application/json; charset=utf-8' }
        expect(last_response.status).to eq(200)
      end
```



If fails reading the body instance:

```
{"interaction_id":"bbcb1f20a4c94e7fb3a3a9cb52b3191f","errors":[{"code":"platform.fault","message":"undefined method `read' for nil","reference":"undefined method `read' for nil,/Users/ildomarg/dev/LoyaltyNZ/hoodoo/lib/hoodoo/services/middleware/middleware.rb:1153:in `monkey_log_inbound_request' | /Users/ildomarg/dev/LoyaltyNZ/hoodoo/lib/hoodoo/services/middleware/middleware.rb:1932:in `process' | /Users/ildomarg/dev/LoyaltyNZ/hoodoo/lib/hoodoo/services/middleware/middleware.rb:661:in `call' | /Users/ildomarg/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/rack-3.1.3/lib/rack/builder.rb:277:in `call' | /Users/ildomarg/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/rack-test-2.1.0/lib/rack/test.rb:360:in `process_request' | /Users/ildomarg/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/rack-test-2.1.0/lib/rack/test.rb:163:in `custom_request' | /Users/ildomarg/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/rack-test-2.1.0/lib/rack/test.rb:112:in `get' | /Users/ildomarg/.rbenv/versions/3.3.0/lib/ruby/3.3.0/forwardable.rb:240:in `get'
```


When performing:

```
body = interaction.rack_request.body.read( MAXIMUM_LOGGED_PAYLOAD_SIZE )
```
`undefined method `read' for nil`




So, what it does says? the behaviour is not the same, body in Rack 2 is not nil, but in Rack 3 it can be nil, even overriding the Request Input

File: lib/hoodoo/services/middleware/middleware.rb (hunk 3 above)
Comment (ildomar-grings) [reply lib/hoodoo/services/middleware/middleware.rb:91]: Just a reformating, following another list above.

File: lib/hoodoo/services/middleware/middleware.rb (hunk 4 above)
Comment (ildomar-grings) [reply lib/hoodoo/services/middleware/middleware.rb:101]: Ditto

File: lib/hoodoo/services/middleware/middleware.rb (hunk 3 above)
Comment (ildomar-grings) [reply lib/hoodoo/services/middleware/middleware.rb]: Reverted

File: lib/hoodoo/services/middleware/middleware.rb (hunk 4 above)
Comment (ildomar-grings) [reply lib/hoodoo/services/middleware/middleware.rb]: Reverted

File: lib/hoodoo/services/middleware/middleware.rb (hunk 2 above)
Comment (ildomar-grings) [reply lib/hoodoo/services/middleware/middleware.rb]: Unbelievable, an check on `rack_request.body.nil?` solves the matter

File: spec/services/services/response_spec.rb (hunk 5 above)
Comment (ildomar-grings) [reply spec/services/services/response_spec.rb:162]: Yes, true.
I see the cascade of problems, so we need to keep this in Upper case. I will find a way to override this.

Comment (ildomar-grings) [reply spec/services/services/response_spec.rb]: Done

File: lib/hoodoo/services/middleware/rack_monkey_patch.rb [hunk 6]
```diff
@@ -71,3 +71,124 @@ def start_and_record_host_and_port( options = nil )
     end
   end
 end
+
```
Comment (davidoram) [inline lib/hoodoo/services/middleware/rack_monkey_patch.rb:74]: I hate these Monkey patches, it seems like it might be a bit fragile

I think is better to accept the new Rack 3.x  behavior for headers, and make this Hoodoo version a major release, with a release not that states users have to change all Headers to only use lower case

File: spec/spec_helper.rb [hunk 7]
```diff
@@ -257,6 +258,7 @@ def spec_helper_start_svc_app_in_thread_for( app_class, use_ssl = false, app_opt
 
   Thread.start do
     app = Rack::Builder.new do
+      use Rack::RewindableInput::Middleware
```
Comment (davidoram) [inline spec/spec_helper.rb:261]: Should we be adding this automatically - say if you include  Hoodoo::Services::Middleware ?

**Review by davidoram — COMMENTED** [review]
Next round of comments

File: lib/hoodoo/services/middleware/rack_monkey_patch.rb (hunk 6 above)
Comment (ildomar-grings) [reply lib/hoodoo/services/middleware/rack_monkey_patch.rb:74]: OK, agreed on the following:
- removing monkey patches, 
- release a hoodoo major version 
- add instructions about how the other services must be adapted to all of these changes

**Review by davidoram — APPROVED** [review]
GTG
